src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Rules
+{
+    class NoRefOrLibFolderInPackageRule : IPackageRule
+    {
+        public string MessageFormat { get; }
+
+        public NoRefOrLibFolderInPackageRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader package)
+        {
+            var files = package.GetFiles();
+            if (!files
+                    .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                    .Any(t => t.StartsWith(PackagingConstants.Folders.Lib + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
```

Comment:
In Release mode, the compiler is hopefully smart enough to precompile the concatenated strings, but just in case you should extract the first parameter of each `StartsWith` call in this file into ` const string`. If the compiler complains that they can't be `const` then it's even more important to make them `private static readonly string`. This avoids needing to concatenate stings for every file in the package and having them available for garbage collection immediately afterwards.